### PR TITLE
Add Filebeat apache access vhost

### PR DESCRIPTION
**Pull request by wixaw**
Hello 

It's a PR for add vhost name in document :)

It's linked with this issue : https://github.com/elastic/beats/issues/4859 

++

### CODE REVIEW

**Comment by wixaw** [comment]
I think my PR has been forgotten? :) ( cc @jsoriano  )

**Comment by jsoriano** [comment]
@wixaw oops, sorry :grimacing: Let me take a look.

File: filebeat/module/apache/access/ingest/default.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
     "grok": {
       "field": "message",
       "patterns":[
+        "%{IPORHOST:http.vhost} %{IPORHOST:source.address} - %{DATA:user.name} \\[%{HTTPDATE:apache.access.time}\\] \"(?:%{WORD:http.request.method} %{DATA:url.original} HTTP/%{NUMBER:http.version}|-)?\" %{NUMBER:http.response.status_code:long} (?:%{NUMBER:http.response.body.bytes:long}|-)( \"%{DATA:http.request.referrer}\")?( \"%{DATA:user_agent.original}\")?",
```
Comment (jsoriano) [inline filebeat/module/apache/access/ingest/default.json]: To follow [ECS](https://www.elastic.co/guide/en/ecs/1.0/ecs-field-reference.html), I think that the vhost should go to the `source.domain`.
And I wonder if in this case we can consider that the source address is always going to be an IP, then we could directly store it in `source.ip`.
Do you know of any case where both fields are domains?

```suggestion
        "%{IPORHOST:source.domain} %{IPORHOST:source.ip} - %{DATA:user.name} \\[%{HTTPDATE:apache.access.time}\\] \"(?:%{WORD:http.request.method} %{DATA:url.original} HTTP/%{NUMBER:http.version}|-)?\" %{NUMBER:http.response.status_code:long} (?:%{NUMBER:http.response.body.bytes:long}|-)( \"%{DATA:http.request.referrer}\")?( \"%{DATA:user_agent.original}\")?",
```

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -208,6 +208,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Add MSSQL module {pull}12079[12079]
 - Add ISO8601 date parsing support for system module. {pull}12568[12568] {pull}12578[12579]
 - Use correct OS path separator in `add_kubernetes_metadata` to support Windows nodes. {pull}9205[9205]
+- Add support for virtual host in Apache access logs {pull}xxxx[xxxxx]
```
Comment (jsoriano) [inline CHANGELOG.next.asciidoc]:
```suggestion
- Add support for virtual host in Apache access logs {pull}12778[12778]
```

File: filebeat/docs/modules/apache.asciidoc [hunk 3]
```diff
@@ -84,3 +84,14 @@ include::../include/var-paths.asciidoc[]
 For a description of each field in the module, see the
 <<exported-fields-apache,exported fields>> section.
 
+=== Virtual Host
+
+See customlog documentation  https://httpd.apache.org/docs/2.4/en/mod/mod_log_config.html
+Add %v config in httpd.conf in log section
```
Comment (jsoriano) [inline filebeat/docs/modules/apache.asciidoc]: I would usually say that we only support default log formats in included pipelines, but I think that it is ok to also support this format as seems to be frequently used, and is included in the official documentation.

File: filebeat/module/apache/access/test/test.log [hunk 4]
```diff
@@ -4,3 +4,4 @@
 172.17.0.1 - - [29/May/2017:19:02:48 +0000] "GET /stringpatch HTTP/1.1" 404 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
 monitoring-server - - [29/May/2017:19:02:48 +0000] "GET /status HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
 127.0.0.1 - - [02/Feb/2019:05:38:45 +0100] "-" 408 152 "-" "-"
+vhost1.domaine.fr 123.123.123.123 - - [04/Jul/2019:09:20:46 +0200] "GET / HTTP/1.1" 200 32953 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36"
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test.log]: As is a different log format, please add it to a new file, something like `test-vhost.log`.

**Comment by jsoriano** [comment]
ok to test

File: filebeat/module/apache/access/ingest/default.json (hunk 1 above)
Comment (wixaw) [reply filebeat/module/apache/access/ingest/default.json]: the second field is never a domain

**Comment by wixaw** [comment]
Hummm 
I'm sorry, I do not understand why Travis and Jenkins growl. Maybe my branch update is broken ?

**Comment by jsoriano** [comment]
> Hummm
> I'm sorry, I do not understand why Travis and Jenkins growl. Maybe my branch update is broken ?

After modifying the documentation you need to run `make update` from the filebeat directory and commit the changes.

The line in CI output mentioning that is this one:
```
06:24:16 Error: some files are not up-to-date. Run 'mage fmt update' then review and commit the changes. Modified: [filebeat/docs/modules/apache.asciidoc]
```

**Comment by wixaw** [comment]
Since your answer I can not do it. I have this same problem : https://discuss.elastic.co/t/getting-error-while-making-custom-beat-panic-failed-determine-libbeat-dir-location/191327

> panic: failed determine libbeat dir location: failed to determine root import path (Did you git init?, Is the project in the GOPATH? GOPATH=/home/wvincent/beats_wix/filebeat, CWD=/home/wvincent/beats_wix/filebeat?): [directory "/home/wvincent/beats_wix/filebeat" is outside source root "/home/wvincent/beats_wix/filebeat/src"]
> 
> goroutine 1 [running]:
> github.com/elastic/beats/dev-tools/mage.LibbeatDir(0xc0000939c0, 0x1, 0x1, 0xc0000d84b0, 0x30)
> 	/local/go/src/github.com/elastic/beats/dev-tools/mage/common.go:729 +0x1d7
> github.com/elastic/beats/dev-tools/mage.init.ializers()
> 	/local/go/src/github.com/elastic/beats/dev-tools/mage/config.go:45 +0x443
> 

I tried to move GOPATH to /home/user/go/ /home/user/beats/ /local/

The mage installation does not work when i start 'make update', i have to do `go install -a github.com/elastic/beats/vendor/github.com/magefile/mage` manualy

I followed the documentation https://www.elastic.co/guide/en/beats/devguide/current/beats-contributing.html

Sorry for not being autonomous about this problem....

**Comment by jsoriano** [comment]
@wixaw did you try to have your beats working copy in `$GOPATH/src/github.com/elastic/beat`?

**Comment by wixaw** [comment]
Hummm i have this error now with real original beats and my data beats 

```
Ϟ>09:34:06wvincent@HOME~/goworkspace/src/github.com/elastic/beats/filebeatʞ make update
mage update
Error determining list of magefiles: failed to list mage gofiles: exit status 1
make: *** [update] Erreur 1
****
```

I thinks it comes from ../libbeat/scripts/Makefile :15

> ES_BEATS?=..## @community_beat Must be set to ./vendor/github.com/elastic/beats. It must always be a relative path.

but is difficult to debug

**Comment by jsoriano** [comment]
@wixaw `make update` uses `mage update` under the hood. You can get more information running it with the `-v` flag, and you can force a rebuild of the targets with `-f`. So try running `mage -f -v update`.

This should give you more information about the error.

If it doesn't, this error seems to appear if the following command fails, can you try to run it?
```
go list -tags=mage -e -f '{{join .GoFiles "||"}}'
```

Also, could you check the `mage` version with `mage --version`?

**Comment by wixaw** [comment]
it's incredible. I went on a other server, and it worked .....
For information here is the error message of my workstation. but do not bother with this. It's in the case or another to the same thing as me

```
Ϟ>09:21:24wvincent@HOME~/go/src/github.com/elastic/beats/filebeatʞ go list -tags=mage -e -f '{{join .GoFiles "||"}}'
unexpected directory layout:
	import path: github.com/magefile/mage/mg
	root: /home/wvincent/go/src
	dir: /home/wvincent/go/src/github.com/elastic/beats/vendor/github.com/magefile/mage/mg
	expand root: /home/wvincent/go/src
	expand dir: /home/wvincent/go/src/github.com/elastic/beats/vendor/github.com/magefile/mage/mg
	separator: /


Ϟ>09:33:18wvincent@HOME~/go/src/github.com/elastic/beatsʞ mage -f -v 
Error determining list of magefiles: failed to list mage gofiles: exit status 1


mage --version
Mage Build Tool v1.8.0


```

**Comment by wixaw** [comment]
ok now, "mage update" remove my news lines of documentation ....  https://github.com/elastic/beats/pull/12778/commits/828600a1be254b78a9c28f805d81f35e26c7db34 
I tried " make docs" but nothing change

**Comment by jsoriano** [comment]
> ok now, "mage update" remove my news lines of documentation .... [828600a](https://github.com/elastic/beats/commit/828600a1be254b78a9c28f805d81f35e26c7db34)
> I tried " make docs" but nothing change

Ok, this is more or less expected, files under ` filebeat/docs/modules` are generated, you need to make the changes in `filebeat/module/apache/_meta/docs.asciidoc` and then run `make update`.

**Comment by wixaw** [comment]
Finally !!!!!
Thank you I would be even more ready for my next PR :stuck_out_tongue_closed_eyes: 
Pray for Travis :pray:

File: filebeat/module/apache/_meta/docs.asciidoc [hunk 5]
```diff
@@ -66,6 +66,19 @@ include::../include/var-paths.asciidoc[]
 
 include::../include/var-paths.asciidoc[]
 
+[float]
+=== Virtual Host
+￼
```
[the remark this answers is not in the document]
Comment (wixaw) [reply filebeat/module/apache/_meta/docs.asciidoc]: humm what? the line break only? or all the doc ?

File: filebeat/module/apache/access/ingest/default.json [hunk 6]
```diff
@@ -1,99 +1,61 @@
...
+      "ignore_missing": true
+    }
+  }],
+  "on_failure" : [{
+    "set" : {
+      "field" : "error.message",
+      "value" : "{{ _ingest.on_failure_message }}"
+    }
+  }]
+}
```
[the remark this answers is not in the document]
Comment (wixaw) [reply filebeat/module/apache/access/ingest/default.json]: I did not understand why everything is rewritten. Do you think it comes from the file format? This is not UTF8?

File: filebeat/module/apache/_meta/docs.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply filebeat/module/apache/_meta/docs.asciidoc]: Doc is great :slightly_smiling_face:

I refer only to this symbol, is it a line break? Why it doesn't appear in other empty lines? Could you try to remove it?

File: filebeat/module/apache/access/ingest/default.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply filebeat/module/apache/access/ingest/default.json]: It seems that the whole file was reformatted, on previous versions opening and closing brackets were in different lines, as in:
```
        },
        {
``` 

In this version opening and closing brackets are in the same line:
```
}, {
```

Also indentation seems to have changed.

I'd prefer if you could keep the previous format, so changes in this PR are limited to the needed ones.

As far as I can see this is not related to UTF8.

**Comment by wixaw** [comment]
Oupss
I'm sorry I do not know why I did not have the good indentation. Maybe because my source was too old?
Here it is corrected.
Thank you Jaime

**Comment by jsoriano** [comment]
Thanks @wixaw, this will need another `make update`.

**Comment by wixaw** [comment]
Thank you for the approval :champagne: 
See you soon for the PR seo parsing of apache: P

**Comment by jsoriano** [comment]
@wixaw test failures in CI seem related to the new test file, could you take a look?

File: filebeat/module/apache/access/test/test-vhost.log [hunk 7]
```diff
@@ -0,0 +1 @@
+vhost1.domaine.fr 192.168.33.2 - - [26/Dec/2016:16:22:14 +0000] "GET /hello HTTP/1.1" 404 499 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:50.0) Gecko/20100101 Firefox/50.0"
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test-vhost.log]: A newline is needed at the end of the file

File: filebeat/module/apache/access/test/test.log [hunk 8]
```diff
@@ -3,4 +3,4 @@
 ::1 - - [26/Dec/2016:16:16:48 +0200] "-" 408 -
 172.17.0.1 - - [29/May/2017:19:02:48 +0000] "GET /stringpatch HTTP/1.1" 404 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
 monitoring-server - - [29/May/2017:19:02:48 +0000] "GET /status HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
-127.0.0.1 - - [02/Feb/2019:05:38:45 +0100] "-" 408 152 "-" "-"
+127.0.0.1 - - [02/Feb/2019:05:38:45 +0100] "-" 408 152 "-" "-"
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test.log]: Here also a new line is needed.

File: filebeat/module/apache/access/test/test-vhost.log-expected.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        "event.dataset": "apache.access",
+        "event.module": "apache",
+        "fileset.name": "access",
+        "http.request.method": "GET",
+        "http.request.referrer": "-",
+        "http.response.body.bytes": 499,
+        "http.response.status_code": 404,
+        "http.version": "1.1",
+        "input.type": "log",
+        "log.offset": 73,
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test-vhost.log-expected.json]:
```suggestion
        "log.offset": 0,
```

File: filebeat/module/apache/access/test/test-vhost.log-expected.json [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        "service.type": "apache",
+        "source.address": "192.168.33.2",
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test-vhost.log-expected.json]:
```suggestion
```

File: filebeat/module/apache/access/test/test-vhost.log-expected.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        "source.domain": "vhost1.domaine.fr",
+        "source.ip": "192.168.33.1",
```
Comment (jsoriano) [inline filebeat/module/apache/access/test/test-vhost.log-expected.json]:
```suggestion
        "source.ip": "192.168.33.2",
```

**Comment by wixaw** [comment]
Thanks, 
It is not easy !
My two clusters (test and production) are in https. I can not find anything in the PR to establish this connection 
 ( https://github.com/elastic/beats/pull/3214/files/d3a4a7eb5271b8096fdf3d691c20627c61ba6521 )

I wanted to use "make start" but i have this error :/ 

```
Step 3/11 : RUN wget https://beats-nightlies.s3.amazonaws.com/filebeat/filebeat-6.0.0-alpha1-SNAPSHOT-linux-x86_64.tar.gz -O filebeat.tar.gz
 ---> Running in f01f3080cd08

--2019-09-04 08:56:29--  https://beats-nightlies.s3.amazonaws.com/filebeat/filebeat-6.0.0-alpha1-SNAPSHOT-linux-x86_64.tar.gz
Resolving beats-nightlies.s3.amazonaws.com (beats-nightlies.s3.amazonaws.com)... 52.216.88.187
Connecting to beats-nightlies.s3.amazonaws.com (beats-nightlies.s3.amazonaws.com)|52.216.88.187|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2019-09-04 08:56:30 ERROR 404: Not Found.
```

**Comment by jsoriano** [comment]
> It is not easy !

Yeah sorry, I agree that this could be more intuitive :grimacing: 

> My two clusters (test and production) are in https. I can not find anything in the PR to establish this connection

Please don't use a real cluster for this, it can remove pipelines and other things, a local elasticsearch is enough, you can start it as you would start any other local elasticsearch.

If you want to use the same version as used in CI, an option to start it is using docker is to run this from the filebeat directory:
```
make start-environment
```
And then use `docker inspect` to get the IP of the elasticsearch container.

**Comment by Verdoso** [comment]
A bit late to the party, but have you guys noticed that the name you are using for the new variable (the virtual host) is not really intuitive?
We will have now
source.domain -> This is not a domain, this the host name (one of them) of the SERVER
source.ip -> The IP of the CLIENT
source.address -> The host name, if available, of the CLIENT
So, all of the variables start with source. but the new one is a value from the server receiving the request, not the client generating the request as the other two. On top of that, the value is a host name, not a domain.
I fear the new name is going to confuse people more often than not.
My 2ec

**Comment by jsoriano** [comment]
@Verdoso thanks, actually this is a very valuable comment.

Regarding storing it as a hostname or as a domain, there have been many discussions about that in ECS, and at the end we use domain for most hostnames, except when identifying a host, that then `host.name` is used.

But it is true that using `source` for one of the server names is confusing

After some internal discussions we think that we should use `destination.domain` for these vhosts.

@wixaw would you like to make the change from `source.domain` to `destination.domain`?

**Comment by Verdoso** [comment]
@jsoriano, thanks to you. If domain is used all over the place to mean hostnames, then at least it is consistent. destination.domain sounds fair enough.
Thanks!

**Comment by wixaw** [comment]
Hello @jsoriano 
I changed the field. :)
However, I still have a problem with the "tests/system/test_modules.py"
All tests are errored and in the list I do not have "test-vhost.log"

Exemple : 

> ======================================================================
> ERROR: test_fileset_file_0_apache (test_modules.Test)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/go/src/github.com/elastic/beats/filebeat/build/python-env/lib/python2.7/site-packages/parameterized/parameterized.py", line 392, in standalone_func
>     return func(*(a + p.args), **p.kwargs)
>   File "/root/go/src/github.com/elastic/beats/filebeat/tests/system/test_modules.py", line 106, in test_fileset_file
>     cfgfile=cfgfile)
>   File "/root/go/src/github.com/elastic/beats/filebeat/tests/system/test_modules.py", line 150, in run_on_file
>     bufsize=0).wait()
>   File "/usr/lib64/python2.7/subprocess.py", line 711, in __init__
>     errread, errwrite)
>   File "/usr/lib64/python2.7/subprocess.py", line 1327, in _execute_child
>     raise child_exception
> OSError: [Errno 2] No such file or directory
> -------------------- >> begin captured stdout << ---------------------
> Using elasticsearch: http://172.18.0.4:9200
> Testing apache/access on /root/go/src/github.com/elastic/beats/filebeat/tests/system/../../module/apache/access/test/darwin-2.4.23.log

**Comment by wixaw** [comment]
I do not understand this
image

**Comment by jsoriano** [comment]
> OSError: [Errno 2] No such file or directory

Oh sorry :man_facepalming: I forgot to mention that you also need a test binary to execute. You can build it with `go test -c`, it will be built as `filebeat.test`.

My `Waiting for green` comment meant to wait for green CI builds before merging :slightly_smiling_face:

**Comment by jsoriano** [comment]
Failures in CI are not related, merging this, thanks!

**Comment by wixaw** [comment]
yeah !!! thanks ! 

See you soon :smile:

File: filebeat/module/apache/access/ingest/default.json (hunk 1 above)
Comment (pa-jberanek) [reply filebeat/module/apache/access/ingest/default.json]: Eh, how is the vhost anything to do with the source? It's an attribute of the **server**...

**Pull request merged**